mist-core/src/main/java/edu/snu/mist/core/task/checkpointing/DefaultCheckpointManagerImpl.java

```diff
@@ -59,101 +59,53 @@
   private final GroupAllocationTableModifier groupAllocationTableModifier;
 
   /**
-   * The query manager for the task.
-   */
-  private final QueryManager queryManager;
+   * The query manager for the task. Use InjectionFuture to avoid cyclic injection.
+   **/
+  private final InjectionFuture<QueryManager> queryManagerFuture;
 
   @Inject
   private DefaultCheckpointManagerImpl(final ApplicationMap applicationMap,
                                        final GroupMap groupMap,
                                        final GroupCheckpointStore groupCheckpointStore,
                                        final GroupAllocationTableModifier groupAllocationTableModifier,
-                                       final QueryManager queryManager) {
+                                       final InjectionFuture<QueryManager> queryManagerFuture) {
     this.applicationMap = applicationMap;
     this.groupMap = groupMap;
     this.checkpointStore = groupCheckpointStore;
     this.groupAllocationTableModifier = groupAllocationTableModifier;
-    this.queryManager = queryManager;
+    this.queryManagerFuture = queryManagerFuture;
+  }
+
+  @Override
+  public boolean storeQuery(final AvroDag avroDag) {
+    return checkpointStore.saveQuery(avroDag);
   }
 
   @Override
   public void recoverGroup(final String groupId) throws IOException {
-    final GroupCheckpoint checkpoint;
+    final Map<String, QueryCheckpoint> queryCheckpointMap;
+    final List<AvroDag> dagList;
+    final QueryManager queryManager = queryManagerFuture.get();
     try {
-      checkpoint = checkpointStore.loadGroupCheckpoint(groupId);
+      // Load the queries.
+      final List<String> queryIdListInGroup = new ArrayList<>();
+      for (final Query query : groupMap.get(groupId).getQueries()) {
```

Comment:
At this point of time when recovery is started, does the `groupMap` have any `Group`s?
I think the `queryIdListInGroup` should be retrived from the `queryCheckpointMap`.

Reply:
@kyungissac It seems you are right :) And I think I finally found out why my test doesn't work! Thanks a lot for your comment.